Change open-amp/libmetal to build non-recursive

galak opened the pull request:
This fixes issue #7673.

galak committed:
ext: open-amp: Change build integration so its not recursive

With recent changes to open-amp we can now include and build it directly
as a zephyr library rather than doing a recursive make.  We remove
ext/lib/ipc/open-amp.cmake as part of this change and introduce a
Kconfig option for open-amp.

Fixes: #7673

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
ext: libmetal: Update import of libmetal

Pull in a slightly more recent version of libmetal with changes
to the build system to improve integration with Zephyr.  This change
will break anything that builds with Zephyr and libmetal (which at this
point is only the open-amp example).  Will fix that shortly.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
ext: libmetal: Change build integration so its not recursive

With recent changes to libmetal we can now include and build it directly
as a zephyr library rather than doing a recursive make.  We remove
ext/hal/libmetal.cmake as part of this change and introduce a Kconfig
option for libmetal.

This is a partial fix for issue #7673.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
ext: open-amp: Update import of open-amp

Pull in a slightly more recent version of open-amp with changes
to the build system to improve integration with Zephyr.  The recent
changes let us embedded open-amp without having to use recursive
building.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

SebastianBoe commented on ext/hal/CMakeLists.txt:
```diff
@@ -1,5 +1,13 @@
+if(CONFIG_LIBMETAL)
```
This libmetal-specific code does not belong in ext/hal/CMakeLists.txt.

It should be pushed into libmetal somehow.

galak replied:
How, I don't think it makes sense to modify the CMakeLists.txt file in libmetal.  I don't see a way to do this beyond moving libmetal down another dir, but that seems silly just to avoid a little CMake foo in ext/hal that is clearly related to libmetal.

nashif replied:
I agree with @SebastianBoe , we should not allow the top level CMakeLists.txt file to be come the place where modules are being configured, this will become unmaintainable if everyone starts putting their logic in there. Going down one level seems reasonable, we did this for a few other modules. Also, the README.libmetal does not belong in there.
All of this seems silly now, but when we move to multi-git structure, it will become necessary.

SebastianBoe replied:
"Going down one level seems reasonable, we did this for a few other modules."
+1

galak replied:
Ok, I'll change it, what's an example where we do this one additional level stuff?

SebastianBoe commented on ext/hal/CMakeLists.txt:
```diff
@@ -1,5 +1,13 @@
+
 add_subdirectory(atmel)
 add_subdirectory(cmsis)
+add_subdirectory_ifdef(CONFIG_LIBMETAL libmetal)
```
_if_kconfig(libmetal) is preferable when the define matches the directory name.

SebastianBoe commented on ext/lib/ipc/CMakeLists.txt:
```diff
@@ -0,0 +1,15 @@
+  set(WITH_LIBMETAL_FIND OFF CACHE BOOL "" FORCE)
+  set(LIBMETAL_INCLUDE_DIR ${ZEPHYR_BINARY_DIR}/ext/hal/libmetal/lib/include)
+  set(LIBMETAL_LIB ${ZEPHYR_BINARY_DIR}/ext/hal/libmetal/lib)
+endif(CONFIG_OPENAMP)
```
This should also be moved into an openamp-specific build script.

SebastianBoe commented:
Shouldn't there be a DNM tag?

It doesn't seem to build ...

SebastianBoe commented:
Okay, so it is broken on master as well.

Anyway, I can't review openamp without testing it. So this review is blocked until #7760 is resolved.

galak commented:
> Anyway, I can't review openamp without testing it. So this review is blocked until #7760 is resolved.

Why can't you test it w/Make instead of ninja?

SebastianBoe commented:
I can. Was not aware #7760 was Ninja-specific.

galak committed:
ext: move libmetal to hal

libmetal itself isn't an ipc library, its a generic HAL abstraction
library so move it into ext/hal where it belongs.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
ext: open-amp: move open-amp down one dir

Move open-amp to be in ext/lib/ipc/open-amp/open-amp.  This allows us to
be Zephyr specific files and config like README (for import),
CMakeLists.txt file, Kconfig, etc in ext/lib/ipc/open-amp/ that don't
conflict with any files that might have the same name in
ext/lib/ipc/open-amp/open-amp.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

SebastianBoe commented:
> what docs do you expect?

A toplevel comment in each source file.

SebastianBoe commented:
Is there any way for a user to know whether OpenAMP is compatible with his platform?

E.g. which platforms do we claim support for the sample? Just lpcxpresso54114_m4? If so that should be explicitly stated in the sample docs instead of implied through the command example and build script code.

The same applies to open-amp itself. Do we claim that open-amp will work on the intersecting platforms of open-amp and Zephyr for instance? If so this should be explicitly documented in open-amp's documentation.

SebastianBoe commented:
For application's with self-contained slaves that don't need to import configuration from the master I believe it will be cleaner to have master Kconfig point to an application source directory instead of an image.

As can be seen in this patch:

https://github.com/SebastianBoe/zephyr/commit/f059efe14d25b7e6b0abf66d5420e6c6bbc779f3

It results in less copy-paste boilerplate in the master application build scripts and less magic paths in the prj.conf.

galak commented:
> Is there any way for a user to know whether OpenAMP is compatible with his platform?
> 
> E.g. which platforms do we claim support for the sample? Just lpcxpresso54114_m4? If so that should be explicitly stated in the sample docs instead of implied through the command example and build script code.
> 
> The same applies to open-amp itself. Do we claim that open-amp will work on all Zephyr platforms, or just the intersecting platforms of open-amp and Zephyr? This should be explicitly documented in open-amp's documentation.

So fair, but is that directly related to these changes?

nashif commented:
> So fair, but is that directly related to these changes?

agree with @galak here, this is about the high priority bug we have lingering, lets resolve any documentation and configuration enhancements to some other PR, it this PR resolve the recursive build issue, lets merge it.

SebastianBoe commented:
> So fair, but is that directly related to these changes?

No, but my review of the Open-AMP integration was dismissed so I am doing it now.
